XenModel/Actions/Folders/FolderAction.cs

```diff
@@ -192,9 +227,68 @@ public override void RecomputeCanCancel()
             // CanCancel is always true.
         }
 
+        private Dictionary<IXenConnection, List<string>> emptyFolders = new Dictionary<IXenConnection, List<string>>();
+
+       /// <summary>
+       /// Mark a folder as empty. The folder will be added to a collection of empty folders that will be used later to update the server
+       /// </summary>
+        private void MarkEmptyFolder(IXenConnection connection, string folder)
+        {
+            if (!emptyFolders.ContainsKey(connection))
+                emptyFolders.Add(connection, new List<string>());
+            emptyFolders[connection].Add(folder);
+        }
+
+        private List<string> nonEmptyFolders = new List<string>();
+
+        /// <summary>
+        /// Mark a folder as non-empty. The folder will be added to a collection of non-empty folders that will be used later to update the server
+        /// </summary>
+        private void MarkNonEmptyFolder(string folder)
+        {
+            nonEmptyFolders.Add(folder);
+        }
+        
+        /// <summary>
+        /// Update the EMPTY_FOLDERS property on all servers, using the emptyFolders and nonEmptyFolders collections
+        /// </summary>
+        private void UpdateEmptyFolders(Func<bool> cancelling)
+        {
+            foreach (var con in emptyFolders.Keys)
+            {
+                AddFoldersToEmptyList(con, cancelling, emptyFolders[con].ToArray());
+            }
+            RemoveFoldersFromEmptyList(nonEmptyFolders.ToArray(), cancelling);
+        }
+
+        internal void DeleteOrMove(List<IXenObject> objects, Folder folder, Func<bool> cancelling)
+        {
+            IXenConnection connection = null;
+            if (objects.Count > 0 && objects[0] != null)

```

Comment:
Is it likely to have null objects in the list? If so we need to add a null check in the other DeleteOrMove overload.
Alternatively we can filter out null objects at the beginning of this method - also return if the list is empty.